### ISSUE TITLE
Proposal: Mark `dart:ui#Paint` `final`.

### ISSUE THREAD

**Issue by matanlurey** [flutter/flutter#142937]
`Paint` (https://api.flutter.dev/flutter/dart-ui/Paint-class.html) is a mutable, framework-specific (i.e. Flutter) "builder" type API.

For example:
```dart
void example() {
  Paint paint = Paint();
  if (paintTheTownRed) {
    paint.color = Colors.red;
  }
  /* ... */
}
```

For historic reasons, `Paint` is "open" (it can be implemented, extended, or I believe even mixed-in).

Dart added [class modifiers](https://dart.dev/language/class-modifiers) in Dart 3.x, which we could utilize:
```diff
- class Paint {
+ final class Paint {
```

Of course this would be a **breaking change** with no trivial workarounds (other than "stop doing that").

## Principles

Here is why I think we should make this change, independent of existing usages:

1. The `Paint` API was never intended to be user-extendable (the implementation is not versioned, for example)
2. As a mutable object, there is no (good) reason to mock or stub the API - just create a "real" object and mutate it
3. A `final` class would let us more easily add new methods (or change signatures over time)

For example https://github.com/flutter/flutter/issues/142871 becomes trivial if the class was `final`.

## Existing Code

1. [Within `org:flutter`: ~0](https://github.com/search?q=%2F%28implements%7Cextends%29%5Cs*Paint%5Cb%2F+language%3ADart+org%3Aflutter&type=code), well [1 archived repository](https://github.com/flutter/photobooth/blob/dba9e669a6744191a19f408a64a93a28bc7dc92c/test/photobooth/widgets/shutter_button_test.dart#L14).
2. [Github-wide: ~14](https://github.com/search?q=%2F%28implements%7Cextends%29%5Cs*Paint%5Cb%2F+language%3ADart&type=code). Breakdown below.
3. Internal to Google: I could not find _any_.

| repository | reference | how easy to fix |
| ---------- | ---------- | --------------- |
| `hydro-sdk` | [`paint.dart`](https://github.com/hydro-sdk/hydro-sdk/blob/00430ce9f58d97452a2876bfc7757c129f352b94/lib/cfr/builtins/libs/dart/ui/paint.dart#L231) | 🔴 hard
| `flutter_json_widgets` | [`paint.freezed.dart`](https://github.com/rodydavis/flutter_json_widgets/blob/3bdfc7d349cbb83192c06a3a1c6d6799a19fcd85/lib/widgets/paint.freezed.dart#L33) | ⚪  false positive (generic bounds)
| `flutter_figma-styles-exporter` | [`fill.dart`](https://github.com/aloisdeniel/flutter_figma-styles-exporter/blob/c4cf8f3383e7099c10890586b30b4c162e0a5109/src/lib/themes/styles/fill.dart#L28) | ⚪  over 5+ years old
| `design_patterns` | [`decorator_mode.dart`](https://github.com/FlutterOpen/design_patterns/blob/93b7059718ac42d9d772205eaf697f81fb8aef07/lib/page/decorator/decorator_mode.dart#L39) | ⚪  over 5+ years old
| `bacum` | [`vacuum_paint.dart`](https://github.com/diezep/bacum/blob/efaa89848e24b815e1e78c2b08f96e40f60da6d0/lib/src/paints/vacuum_paint.dart#L5) | 🟢  appears [~unused](https://github.com/search?q=repo%3Adiezep/bacum%20VacuumPaint&type=code)
| `flutter_painter` | [`painter_test.dart`](https://github.com/omarhurani/flutter_painter/blob/8405829f6a33dd607e520b94f18ce817fcfcd262/test/unit/src/views/painters/painter_test.dart#L197) | 🟢  2+ years old, test could use real implementation.
| `flumemo` | [`pen.dart`](https://github.com/mkizka/flumemo/blob/cfaaacd8400d9cf20c13669da9e2aaa28d171687/lib/models/pen.dart#L44) | ⚪  over 5+ years old and archived.
| `blist` | [`padded_stadium_border_test.dart`](https://github.com/Lukard/blist/blob/238422c6f3e2e45c38c83b247f7974a1ee3e8223/test/xds/style/border/padded_stadium_border_test.dart#L14) | 🟢  test, could just us real implementation.

---

In conclusion: I propose to make this breaking change.

**Comment by matanlurey** [comment]
Given teams have triaged, I'm taking this as a "yes, let's do it".
